Document new features for Buildkite hosted agents GA release (#2786)
buildkite/docs: pages/pipelines/hosted_agents/terminal_access.md

# Hosted agents terminal access

The Buildkite hosted agents feature provides you with _terminal/console access_ to jobs running on hosted agents. This feature is useful in allowing you to:

- Understand what components are installed, as you set up your pipeline.
- Test the behavior of different scripts (because they may not be well-documented).
- Debug issues that are not reproducible in your local environment.

This can be useful when migrating your pipelines across to [queues](/docs/clusters/manage-queues) on Buildkite hosted agents.

## Activate terminal access on hosted agents

> Before you can use the terminal access feature on your Buildkite hosted agents, you'll need to activate this feature for your Buildkite organization, which requires Buildkite organization administrator permissions.

Reviewer: We've changed the default. The feature is enabled for all organizations, but can be disabled within the Organization Settings.

Author: Thanks, I'll repurpose/rewrite this section to be 'deactivate and reactivate terminal access on hosted agents'.

Author: This has now been done and since this feature is now active by default, I've moved this section to the end of the page.

To activate hosted agent terminal access:

1. Select **Settings** in the global navigation to access the [**Organization Settings**](https://buildkite.com/organizations/~/settings) page.
1. Select **Pipelines** > **Settings** to access your organization's [**Pipeline Settings**](https://buildkite.com/organizations/~/pipeline-settings) page.
1. Scroll down to the **Hosted Agents SSH** and select the **Enable SSH** button.
1. Confirm this action by selecting the **Enable Hosted Agents SSH** button.

Terminal access is now granted to all Buildkite hosted agents across all clusters within your Buildkite organization.

> Be aware that users require build permissions on pipelines to use the terminal access feature on the pipelines' jobs.

Reviewer: To use the terminal access feature:
- You must be either an organization admin or a cluster maintainer (in the cluster the job is running within)
- You must have build permissions on pipeline that created the job.

Author: Aha, thanks for the additional info about 'cluster maintainers'.

Author: This is now covered in both sections of the page.

## Use hosted agent terminal access

Once this feature is [activated across your Buildkite organization](#activate-hosted-agent-terminal-access), you can access this terminal access feature from a currently building pipeline, when the job of the relevant step is being built.

> As a pipeline is being built, expand the relevant step and as its job is being built, select its **SSH** button. A new browser window will open with terminal you can use to execute commands to investigate your hosted agent's environment, test script behavior and debug other issues.

Author: If we're going with 'terminal' as a name for this feature, maybe we should change the wording of this button in the UI to:
**Terminal**

# Hosted agents terminal access

The Buildkite hosted agents feature provides you with _terminal/console access_ to jobs running on hosted agents. This feature is useful in allowing you to:

- Understand what components are installed, as you set up your pipeline.
- Test the behavior of different scripts (because they may not be well-documented).
- Debug issues that are not reproducible in your local environment.

This can be useful when migrating your pipelines across to [queues](/docs/clusters/manage-queues) on Buildkite hosted agents.

## Use terminal access on hosted agents

Assuming that [terminal access is active across your Buildkite organization](#deactivate-and-reactivate-terminal-access-on-hosted-agents), you can access this terminal access feature from a currently building pipeline, when the job of the relevant step is being built.

The terminal access feature is available to users who:

- have build permissions on the pipeline that created the job
- are a [maintainer of the cluster](/docs/clusters/manage-clusters#manage-maintainers-on-a-cluster) containing this pipeline

> - are a Buildkite organization administrator of this cluster

Reviewer: One of these can be true to grant access. You must be a maintainer of the cluster OR are a Buildkite organization administrator

Author: OK, if I were to modify the leading line to:

> The terminal access feature is available to users where any of the following is true:

Author: I've made these modifications (assuming these are correct).

> The terminal access feature is available to users who have/are any of the following:
> 
> - build permissions on the pipeline that created the job
> - a [maintainer of the cluster](/docs/clusters/manage-clusters#manage-maintainers-on-a-cluster) containing this pipeline
> - a Buildkite organization administrator of this cluster

Reviewer: ```suggestion
The terminal access feature is available to users who have:

- build permissions on the pipeline that created the job

and one of the following:

- a [maintainer of the cluster](/docs/clusters/manage-clusters#manage-maintainers-on-a-cluster) containing this pipeline
- a Buildkite organization administrator of this cluster
```

Author: Actually, from my own testing, this isn't true...

> build permissions on the pipeline that created the job

I set up a user account _with only this permission_ and I was able to access the terminal.
